File: src/leetcode/2025/11/Solution83.cpp
#include <iostream>
using namespace std;

//  Definition for singly-linked list.
struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *deleteDuplicates(ListNode *head)
    {
        if (!head)
            return nullptr;

        ListNode *current = head;
        while (current->next)
        {
            if (current->val == current->next->val)
            {
                ListNode *temp = current->next;
                current->next = current->next->next;
                delete temp;
            }
            else
            {
                current = current->next;
            }
        }
        return head;
    }
};

int main()
{
    Solution solution;
    // Test case 1: List with duplicates
    ListNode *head1 = new ListNode(1, new ListNode(1, new ListNode(2)));
    ListNode *result1 = solution.deleteDuplicates(head1);
    // Expected Output: 1 -> 2
    while (result1)
    {
        std::cout << result1->val << " ";
        result1 = result1->next;
    }
    std::cout << std::endl;

    // Test case 2: List with multiple duplicates
    ListNode *head2 = new ListNode(1, new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(3)))));
    ListNode *result2 = solution.deleteDuplicates(head2);
    // Expected Output: 1 -> 2 -> 3
    while (result2)
    {
        std::cout << result2->val << " ";
        result2 = result2->next;
    }
    std::cout << std::endl;

    return 0;
}

File: src/leetcode/2025/11/Solution83.java
// Definition for singly-linked list.
class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }

    ListNode(int val, ListNode next) {
        this.val = val;
        this.next = next;
    }

    public String toString() {
        StringBuilder sb = new StringBuilder();
        ListNode current = this;
        while (current != null) {
            sb.append(current.val);
            if (current.next != null) {
                sb.append("->");
            }
            current = current.next;
        }
        return sb.toString();
    }
}

public class Solution83 {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null || head.next == null) {
            return head;
        }
        ListNode current = head;
        while (current.next != null) {
            if (current.val == current.next.val) {
                current.next = current.next.next;
            } else {
                current = current.next;
            }
        }
        return head;
    }

    public static void main(String[] args) {
        Solution83 solution = new Solution83();

        ListNode head1 = new ListNode(1, new ListNode(1, new ListNode(2)));
        ListNode result1 = solution.deleteDuplicates(head1);
        System.out.println(result1); // Output: 1->2

        ListNode head2 = new ListNode(1, new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(3)))));
        ListNode result2 = solution.deleteDuplicates(head2);
        System.out.println(result2); // Output: 1->2->3
    }
}